internal/provider/resource_pipeline.go

```diff
@@ -0,0 +1,366 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/humanitec/humanitec-go-autogen"
+	"github.com/humanitec/humanitec-go-autogen/client"
+)
+
+// Ensure provider defined types fully satisfy framework interfaces
+var _ resource.Resource = &ResourcePipeline{}
+var _ resource.ResourceWithImportState = &ResourcePipeline{}
+
+func NewResourcePipeline() resource.Resource {
+	return &ResourcePipeline{}
+}
+
+// ResourceRule defines the resource implementation.
+type ResourcePipeline struct {
+	client *humanitec.Client
+	orgID  string
+}
+
+func (r *ResourcePipeline) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_pipeline"
+}
+
+func (r *ResourcePipeline) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "An Automation Rule defining how and when artefacts in an environment should be updated.",
+
+		Attributes: map[string]schema.Attribute{
+			"app_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the Application that the Rule should belong to.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"definition": schema.StringAttribute{
+				MarkdownDescription: "The YAML definition of the pipeline",
+				Required:            true,
+			},
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				Computed:            true,
+			},
+			"created_at": schema.StringAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				Computed:            true,
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				Computed:            true,
+			},
+			"status": schema.StringAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				Computed:            true,
+			},
+			"version": schema.StringAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				Computed:            true,
+			},
+			"metadata": schema.MapAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				ElementType: types.StringType,
+				Computed:            true,
+			},
+			"trigger_types": schema.SetAttribute{
+				MarkdownDescription: "The Name of the pipeline",
+				ElementType: types.StringType,
+				Computed:            true,
+			},
+			"etag": schema.StringAttribute{
```

Comment:
I'd suggest removing this from the schema since it may change outside of any updates here. It also doesn't really have much value here.

Reply:
If I remove it from the schema, I'll also need to remove it from the model. This removal will result in the update request lacking the 'etag' field. While it's true that the 'etag' can change in the meantime, this would indicate that other changes have occurred, necessitating a Terraform state refresh

Reply:
Would the etag also change outside terraform? E.g. when the pipeline status changes?

Reply:
Yes it would

Reply:
Any changes to pipeline would result in generating a new etag. However, pipeline is a static resource, there is no dynamic fields that could change over the time.

About pipeline status, it could be either: `active` or `deleting`. The second one means, that pipeline is marked for deletion, but there are some pending runs that has to complete. When all associated runs complete, it will be deleted and it's the most dynamic behavior of the pipeline

Reply:
We could remove it from the model?

Reply:
It's fine to remove the etag from the update - since this TF provider should be the authority on the field - it should override any other update out of band to this pipeline.

Reply:
I think this would be problematic as for example in atlantis a plan file is generated when the PR is opened and only applied after review and there isn't another read in between, so any (unrelated) outside change would fail the apply here 🤔.

Reply:
Convinced! etag removed 😅